plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/firstiteration/FirstIteration.java

```diff
@@ -0,0 +1,341 @@
+            toPlace = rectangles.get(rectangleIdx);
+
+            // determine drawing metrics for different candiate positions/placement options
+            DrawingDataDescriptor lpRight = DrawingDataDescriptor.CANDIDATE_POSITION_LAST_PLACED_RIGHT;
+            opt1 = calcValuesForOpt(lpRight, toPlace, lastPlaced, currentValues, placedRects);
+
+            DrawingDataDescriptor lpBelow = DrawingDataDescriptor.CANDIDATE_POSITION_LAST_PLACED_BELOW;
+            opt2 = calcValuesForOpt(lpBelow, toPlace, lastPlaced, currentValues, placedRects);
+
+            DrawingDataDescriptor wdRight = DrawingDataDescriptor.CANDIDATE_POSITION_WHOLE_DRAWING_RIGHT;
+            opt3 = calcValuesForOpt(wdRight, toPlace, lastPlaced, currentValues, placedRects);
+
+            DrawingDataDescriptor wdBelow = DrawingDataDescriptor.CANDIDATE_POSITION_WHOLE_DRAWING_BELOW;
+            opt4 = calcValuesForOpt(wdBelow, toPlace, lastPlaced, currentValues, placedRects);
+
+            bestOpt = findBestCandidate(opt1, opt2, opt3, opt4, toPlace, lastPlaced);
+
+            toPlace.setY(bestOpt.getNextYcoordinate());
+            toPlace.setX(bestOpt.getNextXcoordinate());
+            bestOpt.setPlaOpt(DrawingDataDescriptor.WHOLE_DRAWING);
+            currentValues = bestOpt;
+            lastPlaced = toPlace;
+            placedRects.add(rectangleIdx, toPlace);
+        }
+
+        return currentValues;
+    }
+
+    // HELPING METHODS
+
+    /**
+     * Determines according to the selected packing strategy, which option out of the given ones is the best.
+     * 
+     * @param opt1
+     *            option lastPlaced right.
+     * @param opt2
+     *            option lastPlaced bottom.
+     * @param opt3
+     *            option complete drawing right.
+     * @param opt4
+     *            option complete drawing below.
+     * @param toPlace
+     *            rectangle to be placed on drawing area.
+     * @param lastPlaced
+     *            rectangle that was placed last on the drawing area.
+     * @return returns the best option out of the given ones or null, if there is no best option found.
+     */
+    private DrawingData findBestCandidate(final DrawingData opt1, final DrawingData opt2, final DrawingData opt3,
+            final DrawingData opt4, final ElkNode toPlace, final ElkNode lastPlaced) {
+        List<DrawingData> candidates = new ArrayList<>();
+        candidates.add(opt1);
+        candidates.add(opt2);
+        candidates.add(opt3);
+        candidates.add(opt4);
+
+        switch (packingStrat) {
+        // examines the candidates according to the packing strategy.
+        case MAX_SCALE_DRIVEN:
+            candidates = BestOptionFinder.findMaxScale(candidates);
+            if (candidates.size() > 1) {
+                candidates = BestOptionFinder.findMinArea(candidates);
+                if (candidates.size() > 1) {
+                    candidates = BestOptionFinder.findBestAspectRatio(candidates, dar);
```

Comment:
The pattern here feels like you want to make the different "BestOptionFinder" functions implementations of an interface (i.e. use the strategy pattern). Then you could add instances of those to a list in the desired order and simply apply them in a single loop (completely avoiding the if three if-cascades here).

Reply:
That one rings a bell, great idea. Cleans up the code ALOT.